r1cs/gadgets/std/src/fields/nonnative/nonnative_field_mul_result_gadget.rs

```diff
@@ -296,15 +340,14 @@ impl<SimulationF: PrimeField, ConstraintF: PrimeField>
         };
 
         // alloc r in normal form. 
-        // Costs `C = len(p)` constraints.
+        // Costs `C = len(p) + num_limbs` constraints.
```

Comment:
I think it was right. r must be a residue modulo p, but for efficiency (checking r < p would be more expensive) we allow it to be < 2^len(p). So allocating it should cost len(p) boolean constraint plus a linear combination.

Reply:
The num_limbs was added because of the "unpacking constraints". Each limb is converted to bit representation with `to_bits_with_length_restriction`, which costs num_bits alloc constraints + 1 unpacking constraints. The 1 was did not take into account in previous estimations, and this partially contributed to the mismatch between the estimation computed in the `find_parameters` function and the actual number of constraints.

Reply:
> The num_limbs is added because of the "unpacking constraints". Each limb is converted to bit representation with `to_bits_with_length_restriction`, which costs num_bits alloc constraints + 1 unpacking constraints. The 1 was did not take into account in previous estimations, and this partially contributed to the mismatch between the estimation computed in the `find_parameters` function and the actual number of constraints.

Could this be done in another way avoiding these extra constraints? I don't see the need of them.

Reply:
Something in my constraint counts was wrong. That was clear from the mismatch in the number of constraints from the optimum finder versus the constraint sythesizer. It seems that I missed to count the linear "packing" constraints in the `alloc()`.